File: NiceReads/Backend/TOS/Common/TOS.Data.MongoDB/Repositories/Repository.cs
using MongoDB.Bson;
using MongoDB.Driver;
using System.Collections.Generic;
using System.Threading.Tasks;
using TOS.Common.MongoDB;
using TOS.Data.Repositories;

namespace TOS.Data.MongoDB.Repositories
{
    public abstract class Repository<TModel> : IRepository<TModel, ObjectId>
        where TModel : IDocumentModel
    {        
        private readonly IMongoDatabase _mongoDatabase;
        private readonly string _collectionName;

        protected Repository(IDatabaseProvider databaseProvider)
        {
            _mongoDatabase = databaseProvider.Database;
            _collectionName = databaseProvider.CollectionNameProvider.GetCollectionName<TModel>();
        }

        protected IMongoCollection<TModel> Collection => _mongoDatabase
                .GetCollection<TModel>(_collectionName);

        public virtual ObjectId Add(TModel model)
        {
            Collection.InsertOne(model);
            return model.Id;
        }

        public virtual async Task<ObjectId> AddAsync(TModel model)
        {
            await Collection.InsertOneAsync(model);
            return model.Id;
        }

        public virtual void AddRange(IEnumerable<TModel> models)
        {
            Collection.InsertMany(models);
        }

        public virtual async Task AddRangeAsync(IEnumerable<TModel> models)
        {
            await Collection.InsertManyAsync(models);
        }

        public virtual bool Delete(ObjectId id)
        {
            return Collection.DeleteOne(m => m.Id == id).DeletedCount == 1;
        }

        public virtual async Task<bool> DeleteAsync(ObjectId id)
        {
            return (await Collection.DeleteOneAsync(m => m.Id == id)).DeletedCount == 1;
        }

        public virtual TModel GetById(ObjectId id)
        {
            return Collection.Find(m => m.Id == id).FirstOrDefault();
        }

        public virtual async Task<TModel> GetByIdAsync(ObjectId id)
        {
            return await Collection.Find(m => m.Id == id).FirstOrDefaultAsync();
        }

        public virtual void Update(TModel model)
        {
            Collection.ReplaceOne(m => m.Id == model.Id, model);
        }

        public virtual async Task UpdateAsync(TModel model)
        {
            await Collection.ReplaceOneAsync(m => m.Id == model.Id, model);
        }
    }
}

File: NiceReads/Backend/Server/tos-server/src/data/repository.ts
import { injectable } from 'inversify';
import { constr } from '../common/common.types';
import { IEntity } from '../common/datamodel/entity';
import { ICollection } from '../database/collection';
import { IDatabase } from '../database/database';

export interface IRepository<TEntity extends IEntity<TId>, TId> {
    add(entity: TEntity): TId;
    addAsync(entity: TEntity): Promise<TId>;
    find(id: TId): TEntity | undefined;
    findAsync(id: TId): Promise<TEntity | undefined>;
    update(entity: TEntity): void;
    updateAsync(entity: TEntity): Promise<void>;
    delete(id: TId): boolean;
    deleteAsync(id: TId): Promise<boolean>;
}

@injectable()
export abstract class Repository<TEntity extends IEntity<TId>, TId> implements IRepository<TEntity, TId> {

    private repositoryCollection: ICollection<TEntity, TId>;

    constructor(protected readonly database: IDatabase<TId>) { }

    protected abstract getEntityType(): constr<TEntity>;

    protected get collection(): ICollection<TEntity, TId> {
        if (!this.repositoryCollection) {
            this.repositoryCollection = this.database.getCollection(this.getEntityType());
        }
        return this.repositoryCollection;
    }

    public add(entity: TEntity): TId {
        return this.collection.add(entity);
    }

    public async addAsync(entity: TEntity): Promise<TId> {
        return await this.collection.addAsync(entity);
    }

    public find(id: TId): TEntity | undefined {
        return this.collection.findById(id);
    }

    public async findAsync(id: TId): Promise<TEntity | undefined> {
        return await this.collection.findByIdAsync(id);
    }

    public update(entity: TEntity): void {
        return this.collection.update(entity);
    }

    public async updateAsync(entity: TEntity): Promise<void> {
        return await this.collection.updateAsync(entity);
    }

    public delete(id: TId): boolean {
        return this.collection.deleteById(id);
    }

    public async deleteAsync(id: TId): Promise<boolean> {
        return await this.collection.deleteByIdAsync(id);
    }
}